README.md
![](https://img.shields.io/circleci/project/github/sbardian/react-fullpage-accordion/develop.svg?style=for-the-badge) ![](https://img.shields.io/coveralls/github/sbardian/react-fullpage-accordion/develop.svg?style=for-the-badge)

# Full page (or set height) fancy React accordion

## After doing [Javascript 30](https://javascript30.com/) by Wes Bos, I decided to make his accordion (from lesson 5) a React component! Enjoy.

### Note: Your React application will need to handle importing `.css` files (e.g. `css-loader` if using `webpack`).

CodeSandbox [Demo](https://codesandbox.io/s/r7v9zlrp6n)

### Install:

npm:

```
npm -i react-fullpage-accordion
```

yarn:

```
yarn add react-fullpage-accordion
```

### Usage:

```
import FullpageAccordion from 'react-fullpage-accordion';
```

### Props:

```
<FullpageAccordion
    items={data}   // Required, array of data objects, see below
    height="500px" // Optional, height of panel container
    textStyle={{ fontSize: '20pt' }} // Optional, any inline css object
/>
```

### Data example:

```
const data = [
    {
    title: "title1",
    itemId: 1,
    top: "Hey",
    middle: "Lets",
    bottom: "Dance",
    imageUrl: "https://source.unsplash.com/gYl-UtwNg_I/1500x1500",
    link: "https://www.google.com",
    },
    {
    title: "title2",
    itemId: 2,
    top: "Give",
    middle: "Take",
    bottom: "Receive",
    imageUrl: "https://source.unsplash.com/rFKUFzjPYiQ/1500x1500",
    link: "https://www.google.com",
    },
    {
    title: "title3",
    itemId: 3,
    top: "Experience",
    middle: "It",
    bottom: "Today",
    imageUrl:
        "https://images.unsplash.com/photo-1465188162913-8fb5709d6d57?ixlib=rb-0.3.5&q=80&fm=jpg&crop=faces&cs=tinysrgb&w=1500&h=1500&fit=crop&s=967e8a713a4e395260793fc8c802901d",
    link: "https://www.google.com",
    },
    {
    title: "title4",
    itemId: 4,
    top: "Give",
    middle: "All",
    bottom: "You Can",
    imageUrl: "https://source.unsplash.com/ITjiVXcwVng/1500x1500",
    link: "https://www.google.com",
    },
    {
    title: "title5",
    itemId: 5,
    top: "Life",
    middle: "In",
    bottom: "Motion",
    imageUrl: "https://source.unsplash.com/3MNzGlQM7qs/1500x1500",
    link: "https://www.google.com",
    }
];
```


CHANGELOG.md
# [v2.0.2](https://github.com/sbardian/react-fullpage-accordion/compare/v2.0.1...v2.0.2)

- Remove base css causing customization problems

# [v2.0.1](https://github.com/sbardian/react-fullpage-accordion/compare/v2.0.0...v2.0.1)

- Add codesandbox link
- Fix bug in height and textStyle types
- Enhanced tests
- Migrate to react Hooks (react/react-dom 16.8.0-alpha.1)
- Add coveralls and circleci

# [v2.0.0](https://github.com/sbardian/react-fullpage-accordion/compare/v0.0.1...v2.0.0)

...


src/Panel.test.js
import React from 'react';
import { render, fireEvent } from 'react-testing-library';
import Panel from './Panel';
import { dataSingle } from '../__mocks__/mockData';

describe('Test Panel', async () => {
  // debug();
  it('Should find one panels div', () => {
    const handleClick = jest.fn();
    const { container, debug, getByText } = render(
      <Panel
        item={dataSingle}
        key={dataSingle.itemId}
        onClick={handleClick}
        activePanel="1"
      >
        <p>{dataSingle.top}</p>
        <p>{dataSingle.middle}</p>
        <p>
          <a
            style={{ zIndex: 20 }}
            href="https://www.google.com"
            alt={dataSingle.title}
          >
            {dataSingle.bottom}
          </a>
        </p>
      </Panel>,
    );
    const firstP = getByText('Hey');
    const secondP = getByText('Lets');
    const thirdP = getByText('Dance');
    expect(firstP).toBeTruthy();
    expect(secondP).toBeTruthy();
    expect(thirdP).toBeTruthy();
    // fireEvent.click(container.getElementsByClassName('panel'));
    // expect(handleClick).toHaveBeenCalledTimes(1);
  });
});


src/index.test.js
import React from 'react';
import { render } from 'react-testing-library';
import FullpageAccordion from './index';
import data from '../__mocks__/mockData';

describe('Test FullpageAccordion for panel container div', async () => {
  const { container, debug } = render(<FullpageAccordion items={data} />);
  it('Should find one panels div', () => {
    const panel = container.getElementsByClassName('panels');
    expect(panel).toHaveLength(1);
  });
  it('Should find five panel elements', () => {
    const panels = container.getElementsByClassName('panel');
    expect(panels).toHaveLength(5);
  });
});


src/Panel.js
import React from 'react';
import PropTypes from 'prop-types';

const Panel = ({ children, height, item, activePanel, onClick }) => (
  <div
    type="button"
    key={item.itemId}
    className={`panel ${
      activePanel == item.itemId ? 'open open-active' : null
    }`}
    style={{
      backgroundImage: `url("${item.imageUrl}")`,
      height: height || null,
    }}
    onClick={onClick}
    onKeyDown={onClick}
    role="button"
    tabIndex={0}
    data-key={item.itemId}
  >
    {children}
  </div>
);

Panel.propTypes = {
  item: PropTypes.shape({
    title: PropTypes.string,
    itemId: PropTypes.number,
    top: PropTypes.string,
    middle: PropTypes.string,
    bottom: PropTypes.string,
    imageUrl: PropTypes.string,
  }).isRequired,
  height: PropTypes.string,
  children: PropTypes.node.isRequired,
  activePanel: PropTypes.number,
  onClick: PropTypes.func.isRequired,
};

Panel.defaultProps = {
  height: null,
  activePanel: null,
};

export default Panel;


src/index.js
/* eslint-disable react/forbid-prop-types */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import 'normalize.css';
import Panel from './Panel';
import './styles.css';

class FullpageAccordion extends Component {
  constructor(props) {
    super(props);
    this.state = {
      activePanel: null,
    };
  }

  handleClick = id => {
    const { activePanel } = this.state;
    const value = id == activePanel ? null : id;
    this.setState({
      activePanel: value,
    });
  };

  render() {
    const { items, height, textStyle } = this.props;
    const { activePanel } = this.state;

    return (
      <div className="panels">
        {items.map(item => (
          <Panel
            key={item.itemId}
            item={item}
            height={height}
            onClick={() => this.handleClick(item.itemId)}
            activePanel={activePanel}
          >
            <p style={{ ...textStyle }}>{item.top}</p>
            <p style={{ ...textStyle }}>{item.middle}</p>
            <p>
              <a
                style={{ zIndex: 20, ...textStyle }}
                href="https://www.google.com"
                alt={item.title}
              >
                {item.bottom}
              </a>
            </p>
          </Panel>
        ))}
      </div>
    );
  }
}

FullpageAccordion.propTypes = {
  items: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.string,
      itemId: PropTypes.number,
      top: PropTypes.string,
      middle: PropTypes.string,
      bottom: PropTypes.string,
      imageUrl: PropTypes.string,
    }),
  ).isRequired,
  height: PropTypes.string,
  textStyle: PropTypes.object,
};

FullpageAccordion.defaultProps = {
  height: null,
  textStyle: {},
};

export default FullpageAccordion;


src/styles.js
const styles = {
  panels: {
    minHeight: '100vh',
    overflow: 'hidden',
    display: 'flex',
    color: 'red',
  },
  panel: {
    background: '#6b0f9c',
    boxShadow: 'inset 0 0 0 5px rgba(255, 255, 255, 0.1)',
    color: 'white',
    textAlign: 'center',
    /* Safari transitionend event.propertyName ::: flex */
    /* Chrome + FF transitionend event.propertyName ::: flex-grow */
    transition: 'font-size 0.7s cubic-bezier(0.61, -0.19, 0.7, -0.11)',
    // 'flex 0.7s cubic-bezier(0.61, -0.19, 0.7, -0.11), background 0.2s',
    fontSize: '20px',
    backgroundSize: 'cover',
    backgroundPosition: 'center',
    display: 'flex',
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
  },
  panel1: {
    backgroundImage: "url('https://source.unsplash.com/gYl-UtwNg_I/1500x1500')",
  },
  panel2: {
    backgroundImage: 'url("https://source.unsplash.com/rFKUFzjPYiQ/1500x1500")',
  },
  panel3: {
    backgroundImage:
      'url("https://images.unsplash.com/photo-1465188162913-8fb5709d6d57?ixlib:rb-0.3.5&q:80&fm:jpg&crop:faces&cs:tinysrgb&w:1500&h:1500&fit:crop&s:967e8a713a4e395260793fc8c802901d")',
  },
  panel4: {
    backgroundImage: 'url("https://source.unsplash.com/ITjiVXcwVng/1500x1500")',
  },
  panel5: {
    backgroundImage: 'url("https://source.unsplash.com/3MNzGlQM7qs/1500x1500")',
  },

  /* Flex Children */
  panelChildren: {
    margin: 0,
    width: '100%',
    transition: 'transform 0.5s',
    flex: '1 0 auto',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  },

  // .panel > *:first-child {
  //   transform: translateY(-100%),
  // }

  // .panel.open-active > *:first-child {
  //   transform: translateY(0),
  // }

  // .panel > *:last-child {
  //   transform: translateY(100%),
  // }

  // .panel.open-active > *:last-child {
  //   transform: translateY(0),
  // }

  // .panel p {
  //   text-transform: uppercase,
  //   font-family: 'Amatic SC', cursive,
  //   text-shadow: 0 0 4px rgba(0, 0, 0, 0.72), 0 0 14px rgba(0, 0, 0, 0.45),
  //   font-size: 2em,
  // }

  // .panel p:nth-child(2) {
  //   font-size: 4em,
  // }

  // .panel.open {
  //   font-size: 40px,
  //   flex: 5,
  // }
};

export default styles;


__mocks__/mockData.js
export default [
  {
    title: 'this is a title',
    itemId: 1,
    top: 'Hey',
    middle: 'Lets',
    bottom: 'Dance',
    imageUrl: 'https://source.unsplash.com/gYl-UtwNg_I/1500x1500',
    link: 'https://www.google.com',
  },
  {
    title: 'this is a title',
    itemId: 2,
    top: 'Give',
    middle: 'Take',
    bottom: 'Receive',
    imageUrl: 'https://source.unsplash.com/rFKUFzjPYiQ/1500x1500',
    link: 'https://www.google.com',
  },
  {
    title: 'this is a title',
    itemId: 3,
    top: 'Experience',
    middle: 'It',
    bottom: 'Today',
    imageUrl:
      'https://images.unsplash.com/photo-1465188162913-8fb5709d6d57?ixlib=rb-0.3.5&q=80&fm=jpg&crop=faces&cs=tinysrgb&w=1500&h=1500&fit=crop&s=967e8a713a4e395260793fc8c802901d',
    link: 'https://www.google.com',
  },
  {
    title: 'this is a title',
    itemId: 4,
    top: 'Give',
    middle: 'All',
    bottom: 'You Can',
    imageUrl: 'https://source.unsplash.com/ITjiVXcwVng/1500x1500',
    link: 'https://www.google.com',
  },
  {
    title: 'this is a title',
    itemId: 5,
    top: 'Life',
    middle: 'In',
    bottom: 'Motion',
    imageUrl: 'https://source.unsplash.com/3MNzGlQM7qs/1500x1500',
    link: 'https://www.google.com',
  },
];

export const dataSingle = {
  title: 'this is a title',
  itemId: 1,
  top: 'Hey',
  middle: 'Lets',
  bottom: 'Dance',
  imageUrl: 'https://source.unsplash.com/gYl-UtwNg_I/1500x1500',
  link: 'https://www.google.com',
};
